Integration of xen-troops patches

svlad-90 opened the pull request:
Integration of the xen-troops patches on top of the Android U-Boot allows running Android using Virito devices within the Xen hypervisor domain.

arminn committed:
Handle verity disable command from Android

Android could update vbmeta during "adb disable-verity" command.
Do not fail due to changes in vbmeta if device is unlocked and
continue to boot using "orange" state.
https://source.android.com/security/verifiedboot/boot-flow#unlocked-devices

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
UPSTREAM: Introduce optional interface parameter to bcb

Interface parameter will give the ability to use bcb on
various block devices. The default logic won't be changed.

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
xen: pvblock: Introduce xenguest_android_arm64 board

* This board configuration will be used for Android
  domain under Xen.
* Enable GPT support, since Android needs it.
* Change boot delay to 3 sec
* Add support of the bootm command
* Introduce bcb based logic
* Use ab_select to choose and support A/B partitions
* Enable AVB 2.0
* Add default dts

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
xen: pvblock: Create a generic option for XENGUEST config

Create a generic option for XENGUEST config, which could
be used by derived xen guest modifications.

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Add xen-specific targets to the BAZEL build

We want to build Android U-boot with Xen-specific configuration. Due to
that we need to define our own u-boot build targets in the Bazel
configuration files.

- Added 'xen-guest-android-virtio_aarch64' target to the Bazel build
configuration
- Added 'xen-guest-android-virtio_aarch64_dist' target to the Bazel
build configuration
- Added 'build.config.xen-guest-android-virtio.aarch64' build
configuration file

Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

Commit:
Fixed deref of the nullptr in boot/image-android.c

In case if there no init_boot_hdr available, the mentioned file was
attempting to dereference the nullptr, which was causing a crash.

This patch is intended to fix the crash by extension of the condition to
check whether init_boot_hdr is a null pointer.

Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

Commit:
xenguest_arm64_android.h: Change CONFIG_EXTRA_ENV_SETTINGS

The 'CONFIG_EXTRA_ENV_SETTINGS' was changed to use Android commands
directly. That allows to follow the Android boot process, throw away
previously used workarounds and simplify the U-Boot commands section
a lot.

Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

lorc commented on cmd/bcb.c:
```diff
@@ -175,12 +182,13 @@ static int __bcb_load(int devnum, const char *partp)
 
 static int do_bcb_load(struct cmd_tbl *cmdtp, int flag, int argc,
-		       char * const argv[])
+			   char * const argv[])
```
Looks like some rebase artifact. Could you please remove unnecessary whitespace changes?

I can't see them in the original commit.

svlad-90 replied:
Hi @lorc, in this case, in the original commit author has mixed spaces and tabs in the source code. It was wrong ( even if it has passed a review ). Tabs are usually used in U-Boot. I've fixed that.

svlad-90 replied:
It was agreed, that we should separate formatting change from functional changes. Thus, wither move it to a separate commit after this one, or remove this change at all.

svlad-90 replied:
Hi @lorc ,

Replacement of spaces with tabs was removed.

Done.

lorc commented on include/configs/xenguest_arm64.h:
```diff
@@ -9,14 +10,42 @@
+#endif
 
-#endif /* __XENGUEST_ARM64_H */
+#endif /* __XENGUEST_ARM64_H */
```
Missing EOL at EOF. Looks like this is unnecessary change.

svlad-90 replied:
Hi @lorc, done.

lorc commented on arch/arm/dts/xenguest-android-arm64.dts:
```diff
@@ -0,0 +1,15 @@
+ * Build instructions at xenguest_arm64.rst are inadequate for obtaining a real
+ * devicetree.
+ *
+ * Copyright 2021 Google LLC
```
EPAM 2023?

svlad-90 replied:
Hi @lorc, done.

lorc commented on xen-guest-android-virtio.fragment:
```diff
@@ -0,0 +1,3 @@
+# CONFIG_SANDBOX is not set
+# CONFIG_POWER is not set
+CONFIG_SYS_LOAD_ADDR=0x40000000
```
Missing EOL at EOF. There should be some setting in your editor to automatically add these.

svlad-90 replied:
Hi @lorc, done.

Also, thanks for the hint. I've adjusted the settings in Sublime to add a new line on the "Save" operation.

lorc commented on arch/arm/cpu/armv8/start.S:
```diff
@@ -16,19 +16,27 @@
-.globl	_start
-_start:
 #if defined(CONFIG_LINUX_KERNEL_IMAGE_HEADER)
+
```
Extra whitespace changes all around this patch. It makes review process difficult. 

Please refrain from changes that are not related to intended patch purpose.

svlad-90 replied:
Hi @lorc, I guess you mean "extra new lines", and not the extra "white spaces".

I've removed them. Done.

But, in general, from my point of view, it makes the code more readable in this case. I still believe in this. But you are the maintainer, so I obey. ))

lorc commented on arch/arm/cpu/armv8/start.S:
```diff
@@ -16,19 +16,27 @@
  * use it here.
  */
+
+.globl	_start
```
`_start:` missing? I checked couple of "boot0.h" files, none of them define `_start` symbol

svlad-90 replied:
Hi @lorc, I've checked them all.

|File link|Whether it has _start defined?|
|---|---|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-imx8/boot0.h|No|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-px30/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-sunxi/boot0.h|No|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3036/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3128/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3188/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk322x/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3288/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3308/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3328/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3368/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3399/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rv1108/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-bcm235xx/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-bcm281xx/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/mach-rmobile/include/mach/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/mach-socfpga/include/mach/boot0.h|Yes|

For the majority of cases, there would be no "_start:" defined for ARM64. So, it seems, that we should add "_start:" here. I'll do this.

svlad-90 replied:
Hi @lorc, done.

lorc replied:
Thanks. But why this change is in "Add xen-specific targets to the BAZEL build" commit?

For me it looks like something that should be in a separate commit, probably worth upstreaming.

svlad-90 replied:
Hi @lorc, it is in the same commit, because the Bazel build is failing without this change. As we agreed - each commit should be compilable. That motivated me to place it in such a way. The thing I can do is place the commit with the compilation issue fix before the "Add xen-specific targets to the BAZEL build". That will ensure atomic, still buildable commits. I will do that tomorrow.

lorc replied:
Yep, this is the right thing to do.

svlad-90 replied:
Hi @lorc, done.

lorc reviewed (commented):
The purpose of rebasing patches is not to resolve merge commits, but to produce valid codebase.  Looking at this PR, I can say that it wasn't accomplished: you are adding a bunch of patches that are fixing original rebased patches.  This is incorrect and provides a broken git history.  For example, if I checkout at "Add xen-specific targets to the BAZEL build" and try build, it fails with an error. 

You need to review your changes and squash appropriate fixes into original commits by Andrii to provide complete working history of changes.

Please remember, that no patch should break the build.

svlad-90 commented:
> The purpose of rebasing patches is not to resolve merge commits, but to produce valid codebase. Looking at this PR, I can say that it wasn't accomplished: you are adding a bunch of patches that are fixing original rebased patches. This is incorrect and provides a broken git history. For example, if I checkout at "Add xen-specific targets to the BAZEL build" and try build, it fails with an error.
> 
> You need to review your changes and squash appropriate fixes into original commits by Andrii to provide complete working history of changes.
> 
> Please remember, that no patch should break the build.

Hi @lorc. I think I've got your point. I'll try to follow the provided instructions. I guess, it will take me several hours to finish this activity. I'll let you know once it is finished with an additional comment.

svlad-90 commented:
Hi @lorc, all changes are done.

PR now consists of 8 commits instead of ~13 at the beginning. Most of the changes from the 5 disappeared commits were squashed into Andrii's commits.

Each separate commit was checked to be built successfully with:

- xenguest_android_arm64_defconfig
- xenguest_arm64_defconfig
- Bazel build with 'xen-guest-android-virtio_aarch64_dist' target ( last 3 commits, where this is applicable )

It is confirmed that Android is still up and running after this PR.

Please, have a look at the updated PR.

lorc commented on configs/xenguest_arm64_defconfig:
```diff
@@ -42,3 +42,36 @@ CONFIG_CMD_FAT=y
 # CONFIG_MMC is not set
 # CONFIG_REQUIRE_SERIAL_CONSOLE is not set
 CONFIG_DM_SERIAL=y
+# CONFIG_SYS_SDRAM_BASE is not set
```
Thanks. I have (probably) no objections regarding commit diff, but I have comment about the commit message. It is unclear why you are doing those changes.  I want you to read section "Describe your changes": 

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes

Yes, this is the Linux manual, but it describes requirements that are used in virtually all embedded open source projects. You can ignore the Linux-related parts, but please stick to all other rules. 

For this particular commit message, you also need to drop "pvblock" word, as it is misleading: there is nothing pvblock-specific in this commit. Also, it is not clear what exactly you are trying to accomplish, so it is hard to review this change properly.

svlad-90 replied:
Hi @lorc,

The original commit which introduced almost all those flags is this one:
https://github.com/xen-troops/u-boot/commit/91a77926c6f3ced01c2c2bcf926dd4c8e10136cd
It was upstreamed.

But this file has a big history in upstream, which excludes the majority of the flags commit after commit:
https://github.com/xen-troops/u-boot/commits/xenvm-trout-main/include/configs/xenguest_arm64.h

I'll check which of the flags are still relevant and will remove all the others.

svlad-90 replied:
Hi @lorc,

After the review of what was upstreamed, I've figured out, that I need no any of those flags at all. This patch of Andrii was upstreamed, and then further patches in the upstream moved almost everything to:

- KConfig default values
- xenguest_arm64_defconfig ( yes, some of my added flags were already there )

Done.

P.S. I will review the Android-specific flags from the same perspective. Those flags which would be automatically added through KConfig are not needed to be presented in the defconfig.

svlad-90 replied:
The same flags for xenguest_android_arm64_defconfig were reviewed and changed.

lorc commented on common/avb_pubkey.c:
```diff
@@ -0,0 +1,71 @@
+const char avb_pubkey[] =
```
Whoa!

Should it be auto generated based on keys used? It is quite strange to include some public key, in the "common" section without providing instructions how to obtain private part of it. 

In my understanding this should be something deeply product-specific, but I can be wrong there.

lorc replied:
Actually, I can see that "common/Makefile" issues `bin2c` tool to generate this file. Why it is included in this commit?

svlad-90 replied:
Hi @lorc, the file was added accidentally. I've removed it.

Done.

lorc reviewed (changes requested):
Thanks for the changes. It is in a much better shape now. But there are still couple of relatively minor issues to fix.

Commit:
start.S: Fix compilation issue

Approach with Bazel and repo, which is used by Google to compile U-Boot
is using a clang compiler instead of gcc. That compiler treats
redefinition of the '_start' symbol as error.

This patch is intended to fix a clang-specific compilation error. The
'asm/boot0-linux-kernel-header.h' header defines '_start" symbol by
itself. Thus, code should be slightly changed to avoid symbol
redefinition in case of 'CONFIG_LINUX_KERNEL_IMAGE_HEADER' flag is
defined.

Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

svlad-90 commented:
Hi @lorc, all changes were done. Please have another look at it.

svlad-90 commented:
Hi @lorc, please, do not merge yet. Let me check that system is up and running, after all the today's changes.

lorc reviewed (approved):
Thanks. It is in a much better shape now.

One last comment: Patch "Add xen-specific targets to the BAZEL build" has outdated description, as it mentions "- Fixed comilation issue with redefinition of the '_start' symbol
( clang specific )". Please fix the description and add my r-b tag to all commits:

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

svlad-90 commented:
Hi @lorc, the changes were tested on the target. U-Boot still works fine for booting Android on our virtio set-up.

Regarding r-b and changing the specific commit message - ok. I'll do that and will let you know.

svlad-90 commented:
Hi @lorc, the changes are done.

Thanks for the review! ))

@arminn, @lorc, please, merge the change.

lorc commented:
@arminn, the honor is yours :)
